src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs

```diff
@@ -0,0 +1,285 @@
+        private const byte ATYP_IPV4 = 1;
+        private const byte ATYP_DOMAIN_NAME = 3;
+        private const byte ATYP_IPV6 = 4;
+        private const byte REP_SUCCESS = 0;
+        // private const byte REP_FAILURE = 1;
+        // private const byte REP_NOT_ALLOWED = 2;
+        // private const byte REP_NETWORK_UNREACHABLE = 3;
+        // private const byte REP_HOST_UNREACHABLE = 4;
+        // private const byte REP_CONNECTION_REFUSED = 5;
+        // private const byte REP_TTL_EXPIRED = 6;
+        // private const byte REP_CMD_NOT_SUPPORT = 7;
+        // private const byte REP_ATYP_NOT_SUPPORT = 8;
+        private const byte CD_SUCCESS = 90;
+
+        public static ValueTask EstablishSocksTunnelAsync(Stream stream, string host, int port, Uri proxyUri, ICredentials? proxyCredentials, CancellationToken cancellationToken)
+        {
+            if (string.Equals(proxyUri.Scheme, "socks5", StringComparison.OrdinalIgnoreCase))
+            {
+                return EstablishSocks5TunnelAsync(stream, host, port, proxyUri, proxyCredentials, cancellationToken);
+            }
+            else if (string.Equals(proxyUri.Scheme, "socks4a", StringComparison.OrdinalIgnoreCase))
+            {
+                return EstablishSocks4TunnelAsync(stream, true, host, port, proxyUri, proxyCredentials, cancellationToken);
+            }
+            else if (string.Equals(proxyUri.Scheme, "socks4", StringComparison.OrdinalIgnoreCase))
+            {
+                return EstablishSocks4TunnelAsync(stream, false, host, port, proxyUri, proxyCredentials, cancellationToken);
+            }
+            else
+            {
+                Debug.Fail("Bad socks version.");
+                return default;
+            }
+        }
+
+        private static async ValueTask EstablishSocks5TunnelAsync(Stream stream, string host, int port, Uri proxyUri, ICredentials? proxyCredentials, CancellationToken cancellationToken)
+        {
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(BufferSize);
+
+            try
+            {
+                // https://tools.ietf.org/html/rfc1928
+
+                // +----+----------+----------+
+                // |VER | NMETHODS | METHODS  |
+                // +----+----------+----------+
+                // | 1  |    1     | 1 to 255 |
+                // +----+----------+----------+
+                buffer[0] = ProtocolVersion5;
+                var credentials = proxyCredentials?.GetCredential(proxyUri, "");
+                if (credentials != null)
+                {
+                    buffer[1] = 1;
+                    buffer[2] = METHOD_NO_AUTH;
+                }
+                else
+                {
+                    buffer[1] = 2;
+                    buffer[2] = METHOD_NO_AUTH;
+                    buffer[3] = METHOD_USERNAME_PASSWORD;
+                }
+                await stream.WriteAsync(buffer.AsMemory(0, buffer[1] + 2), cancellationToken).ConfigureAwait(false);
```

Comment:
We should honor the `async` flag on all these writes/reads to call into sync methods if requested (can use a simple helper method).